torch/_dynamo/guards.py

```diff
@@ -675,6 +678,27 @@ def get_sources(t_id, dim):
         for shape_guard in guards:
             self._produce_guard_code(guard, [shape_guard], shape_env=True)
 
+        def _propagate_guard_to_outer_shape_env(guards):
+            out_shape_env_guard_exprs = [
+                guard_expr
+                for guard_expr in guards
+                if "convert_symbool_to_int_with_hint" in guard_expr
+            ]
+
+            def _convert_symbool_to_int_guarded(symbool):
+                return int(symbool)
+
+            overrided_interpreter = {
+                **SYMPY_INTERP,
+                "convert_symbool_to_int_with_hint": _convert_symbool_to_int_guarded,
+            }
+            assert all(
+                eval(guard_expr, self.scope, overrided_interpreter)
+                for guard_expr in out_shape_env_guard_exprs
+            )
+
+        _propagate_guard_to_outer_shape_env(guards)
```

Comment:
It's not clear to me why you need this. Is the problem you don't actually run the guards on the first compilation of the function?

Reply:
🤔 Can we just run the guards unconditionally on the first compilation of a function

Reply:
yes! The guards are only checked at second-time run.